libcudacxx/include/cuda/__memory/align_down.h

```diff
@@ -0,0 +1,51 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of libcu++, the C++ Standard Library for your entire system,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDA___MEMORY_ALIGN_DOWN_H
+#define _CUDA___MEMORY_ALIGN_DOWN_H
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <cuda/__cmath/pow2.h>
+#include <cuda/std/__type_traits/is_pointer.h>
+#include <cuda/std/cstddef>
+#include <cuda/std/cstdint>
+
+#include <cuda/std/__cccl/prologue.h>
+
+_LIBCUDACXX_BEGIN_NAMESPACE_CUDA
+
+template <typename _Tp>
+[[nodiscard]] _CCCL_API _CUDA_VSTD::enable_if_t<_CUDA_VSTD::is_pointer_v<_Tp>, _Tp>
+align_up(_Tp __ptr, size_t __alignment) noexcept
```

Comment:
```suggestion
[[nodiscard]] _CCCL_API _Tp* align_up(_Tp* __ptr, size_t __alignment) noexcept
```
or is there a problem I don't see?

Reply:
Also I believe we should use `_CUDA_VSTD::size_t` and `_CUDA_VSTD::uintptr_t`

Reply:
I used `_Tp` instead of `_Tp*` because I would like to extend it to `annotated_ptr` in the future.